snapcraft/tests/repo/test_snaps.py

```diff
@@ -0,0 +1,365 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+import json
+import os
+import socketserver
+import subprocess
+import tempfile
+from http.server import BaseHTTPRequestHandler
+from threading import Thread
+from unittest import mock
+from urllib import parse
+
+import fixtures
+from testtools.matchers import Equals, Is
+
+from snapcraft import tests
+from snapcraft.internal.repo import errors, snaps
+
+
+class FakeSnapCommand(fixtures.Fixture):
+
+    def __init__(self):
+        self.calls = []
+        self.install_success = True
+        self.refresh_success = True
+
+    def _setUp(self):
+        original_check_call = snaps.check_call
+        original_check_output = snaps.check_output
+
+        def side_effect_check_call(cmd, *args, **kwargs):
+            return side_effect(original_check_call, cmd, *args, **kwargs)
+
+        def side_effect_check_output(cmd, *args, **kwargs):
+            return side_effect(original_check_output, cmd, *args, **kwargs)
+
+        def side_effect(original, cmd, *args, **kwargs):
+            if self._is_snap_command(cmd):
+                self.calls.append(cmd)
+                return self._fake_snap_command(cmd, *args, **kwargs)
+            else:
+                return original(cmd, *args, **kwargs)
+
+        check_call_patcher = mock.patch(
+            'snapcraft.internal.repo.snaps.check_call',
+            side_effect=side_effect_check_call)
+        self.mock_call = check_call_patcher.start()
+        self.addCleanup(check_call_patcher.stop)
+
+        check_output_patcher = mock.patch(
+            'snapcraft.internal.repo.snaps.check_output',
+            side_effect=side_effect_check_output)
+        self.mock_call = check_output_patcher.start()
+        self.addCleanup(check_output_patcher.stop)
+
+    def _get_snap_cmd(self, snap_cmd):
+        try:
+            snap_cmd_index = snap_cmd.index('snap')
+        except ValueError:
+            return ''
+
+        try:
+            return snap_cmd[snap_cmd_index+1]
+        except IndexError:
+            return ''
+
+    def _is_snap_command(self, cmd):
+        return self._get_snap_cmd(cmd) in ['install', 'refresh', 'whoami']
+
+    def _fake_snap_command(self, cmd, *args, **kwargs):
+        cmd = self._get_snap_cmd(cmd)
+        if cmd == 'install' and not self.install_success:
+            raise subprocess.CalledProcessError(returncode=1, cmd=cmd)
+        if cmd == 'refresh' and not self.refresh_success:
+            raise subprocess.CalledProcessError(returncode=1, cmd=cmd)
+        if cmd == 'whoami':
+            return 'email: -'.encode()
+
+
+class FakeSnapdServer(BaseHTTPRequestHandler):
```

Comment:
Having @chipaca look this fake over to make sure everything is sane would be helpful.

Reply:
I sort of asked him the necessary questions about API and showed him the original implementation so I don't think he is needed her @kyrofa

Reply:
can confirm, was sort of asked.